### PR TITLE
doc: clarify config-relative paths for `--config <path>`

### LINKED ISSUE #10991: Document that cargo --config <filename> has unintuitive behavior for config-relative paths
### Problem

Hi there!

I'm currently looking at configs defined by filenames, and it looks like `cargo --config <filename>` defines config-relative paths at one level *above* the config.toml file, rather than at the same level as the config.toml file.

### Steps

For example, if `foo.toml` is created at the root of the workspace:

```
[build]
target-dir = "from-foo-toml"
```

then when you run `cargo build --config foo.toml`, then "from-foo-toml" is created one level above

While this behavior makes sense for files in `.cargo`, this is extremely surprising behavior for files defined over the command-line. I actually think it is OK to do this, but I didn't see any documentation regarding this on https://doc.rust-lang.org/cargo/reference/config.html.

### Possible Solution(s)

Document this on https://doc.rust-lang.org/cargo/reference/config.html and recommend that Cargo  config files be created in a directory like `.cargo/foo.toml`, not `foo.toml`. (A change to this behavior would be a breaking change, so I don't think that's what should happen.)

### Notes

_No response_

### Version

```text
% cargo version --verbose
cargo 1.63.0 (fd9c4297c 2022-07-01)
release: 1.63.0
commit-hash: fd9c4297ccbee36d39e9a79067edab0b614edb5a
commit-date: 2022-07-01
host: x86_64-unknown-linux-gnu
libgit2: 1.4.2 (sys:0.14.2 vendored)
libcurl: 7.83.1-DEV (sys:0.4.55+curl-7.83.1 vendored ssl:OpenSSL/1.1.1n)
os: Pop!_OS 22.04 (jammy) [64-bit]
```

### PR DESCRIPTION
**Pull request by weihanglo**
### What does this PR try to resolve?

Resolves #10991. 

Instead of changing the behaviour (it might break things), we document the current one as @sunshowers [suggesting](https://github.com/rust-lang/cargo/issues/10991#issuecomment-1215931242).

### How should we test and review this PR?

Check changes in 

- `src/doc/src/reference/config.md` and 
- `src/doc/man/includes/section-options-common.md`.

```shell
cargo help build
# Proofread the manpage change.
mdbook serve
# open http://[::1]:3000/reference/config.html#config-relative-paths
# and proofread the config-relative paths section
```

### CODE REVIEW

File: src/doc/src/reference/config.md [hunk 1]
```diff
@@ -238,13 +238,27 @@ precedence rules as other options specified directly with `--config`.
...
-config files, that is relative to the parent directory of the `.cargo`
-directory where the value was defined. For environment variables it is
-relative to the current working directory.
+Paths in config files may be absolute, relative, or a bare name without any path separators.
+Paths for executables without a path separator will use the `PATH` environment variable to search for the executable. 
+Paths for non-executables will be relative to where the config value is defined.
+In particular, rules are:
+
+* For environment variables it is relative to the current working directory.
+* For config files, that is relative to the parent directory of the directory where the value was defined.
```
Comment (epage) [inline src/doc/src/reference/config.md:247]: > For config files, that is relative to

should probably be

> For config files, they are relative to

Comment (epage) [inline src/doc/src/reference/config.md:247]: Should we also explicitly call out `--config key=value`?

File: src/doc/src/reference/config.md [hunk 2]
```diff
@@ -238,13 +238,27 @@ precedence rules as other options specified directly with `--config`.
...
-path separators. Paths for executables without a path separator will use the
-`PATH` environment variable to search for the executable. Paths for
-non-executables will be relative to where the config value is defined. For
-config files, that is relative to the parent directory of the `.cargo`
-directory where the value was defined. For environment variables it is
-relative to the current working directory.
+Paths in config files may be absolute, relative, or a bare name without any path separators.
+Paths for executables without a path separator will use the `PATH` environment variable to search for the executable. 
+Paths for non-executables will be relative to where the config value is defined.
+In particular, rules are:
```
Comment (epage) [inline src/doc/src/reference/config.md:244]: Should we have a newline separating these?

**Comment by weihanglo** [comment]
All issues are resolved. Does it look better?

File: src/doc/src/reference/config.md (hunk 1 above)
Comment (epage) [reply src/doc/src/reference/config.md]: I'm not seeing the change for this

Comment (weihanglo) [reply src/doc/src/reference/config.md]: I thought you were suggesting something like this, no?

https://github.com/rust-lang/cargo/pull/11079/files#diff-0917d2867ed0f7116151e8c28138a91760a17f3fca809cc72523053c2762f3afR248-R250

Comment (epage) [reply src/doc/src/reference/config.md]: I feel like I'm missing something.

I see `config.md` covering
- Environment variables
- Implicit config files
- Explicit config files via `--config <path>`

I'm not seeing `--config key=value` being discussed for where paths would be relative to.

**Review by epage — APPROVED** [review]

File: src/doc/src/reference/config.md (hunk 1 above)
Comment (weihanglo) [reply src/doc/src/reference/config.md]: Current implementation as a reference.
https://github.com/rust-lang/cargo/blob/c1871146e3f5f5f5913a3ef93ca176b062d7ff3e/src/cargo/util/config/value.rs#L66-L75

**Pull request merged**